assignment-1/pkg/controller/controller.go

```diff
@@ -0,0 +1,206 @@
+		context.Background(), GetSnapshotName(pvc), metav1.GetOptions{})
+	if err == nil {
+		return err
+	}
+	return c.createSnapshot(pvc, snapshot)
+}
+
+func (c *Controller) createSnapshot(pvc *corev1.PersistentVolumeClaim, snapshot *v1alpha1.Snapshot) error {
+	snapshotName := GetSnapshotName(pvc)
+	snapshotClassname := GetSnapshotClassname()
+	snapTemplate := &volumesnapshotv1.VolumeSnapshot{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: snapshotName,
+		},
+		Spec: volumesnapshotv1.VolumeSnapshotSpec{
+			VolumeSnapshotClassName: &snapshotClassname,
+			Source: volumesnapshotv1.VolumeSnapshotSource{
+				PersistentVolumeClaimName: &pvc.Name,
+			},
+		},
+	}
+	_, err := c.ExternalSnapshotClient.SnapshotV1().VolumeSnapshots(pvc.Namespace).Create(
+		context.Background(), snapTemplate, metav1.CreateOptions{})
+	if err != nil {
+		return err
+	}
+	err = c.updateStatus(snapshotName, pvc, snapshot)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("snapshot %s created\n", snapshotName)
+	return nil
+}
+
+func (c *Controller) updateStatus(snapshotName string, pvc *corev1.PersistentVolumeClaim, snapshot *v1alpha1.Snapshot) error {
+	snapshotDeepcopy := snapshot.DeepCopy()
+	snapshotDeepcopy.Status.VolumeSnapshotName = snapshotName
+	snapshotDeepcopy.Status.Storage = pvc.Spec.Resources.Requests.Storage().String()
+	snapshotDeepcopy.Status.StorageClassname = *pvc.Spec.StorageClassName
+	_, err := c.SnapshotClientSet.ShahpratikrV1alpha1().Snapshots(pvc.Namespace).Update(
+		context.Background(), snapshotDeepcopy, metav1.UpdateOptions{})
+	return err
+}
+
+func (c *Controller) restoreBackup(snapshot *v1alpha1.Snapshot) error {
+	// Volume snapshot doesn't exists, return error
+	volumeSnapshot, err := c.ExternalSnapshotClient.SnapshotV1().VolumeSnapshots(snapshot.Namespace).Get(
+		context.Background(), snapshot.Status.VolumeSnapshotName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	return c.createPVC(volumeSnapshot, snapshot)
+}
+
+func (c *Controller) createPVC(volumeSnapshot *volumesnapshotv1.VolumeSnapshot, snapshot *v1alpha1.Snapshot) error {
+	apiGroup := GetApiGroup()
+	pvc := corev1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: *volumeSnapshot.Spec.Source.PersistentVolumeClaimName,
+		},
+		Spec: corev1.PersistentVolumeClaimSpec{
+			StorageClassName: &snapshot.Status.StorageClassname,
+			DataSource: &corev1.TypedLocalObjectReference{
+				Name:     volumeSnapshot.Name,
+				Kind:     "VolumeSnapshot",
+				APIGroup: &apiGroup,
+			},
+			AccessModes: []corev1.PersistentVolumeAccessMode{
+				corev1.ReadWriteOnce,
+			},
+			Resources: corev1.ResourceRequirements{
+				Requests: corev1.ResourceList{
+					corev1.ResourceStorage: resource.MustParse(snapshot.Status.Storage),
+				},
+			},
+		},
+	}
+	_, err := c.KubernetesClientSet.CoreV1().PersistentVolumeClaims(
+		volumeSnapshot.Namespace).Create(context.Background(), &pvc, metav1.CreateOptions{})
```

Comment:
instead of creating context.Background(), derive it from parent as arg to the function